core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -418,9 +464,6 @@ final class ShuffleBlockFetcherIterator(
     } else {
       blocks
     }
-    // update metrics
-    numBlocksToFetch += result.size
```

Comment:
when we put the merged blocks back to the input list, it's kind of rollback this method. I think it's better to not update mutable stats here if it can be rollbacked.